Mathlib/Analysis/Convex/EGauge.lean

```diff
@@ -171,24 +192,109 @@ lemma egauge_smul_right (h : c = 0 → s.Nonempty) (x : E) :
-  · simpa only [add_comm] using this hV hU b a y hy hyc x hx hxc (le_of_not_le hxy)
-  refine ⟨y, ?_, hyc⟩
-  rw [smul_add]
-  exact add_mem_add (hU.smul_mono hxy hx) hy
-
-end VectorSpace
+  rw [← egauge_prod_mk hU hV a b, ← add_image_prod]
+  exact MapsTo.egauge_le 𝕜 (LinearMap.fst 𝕜 E E + LinearMap.snd 𝕜 E E) (mapsTo_image _ _) (a, b)
+
+end NormedDivisionRing
+
+section Pi
+
+variable {𝕜 : Type*} [NormedDivisionRing 𝕜]
+  {ι : Type*} {E : ι → Type*} [∀ i, AddCommGroup (E i)] [∀ i, Module 𝕜 (E i)]
+
+/-- The extended gauge of a point `x` in an indexed product
+with respect to a product of finitely many balanced sets `U i`, `i ∈ I`,
+(and the whole spaces for the other indices)
+is the supremum of the extended gauges of the components of `x`
+with respect to the corresponding balanced set.
+
+This version assumes the following technical condition:
+- either `I` is the universal set;
+- or one of `x i`, `i ∈ I`, is nonzero;
+- or `𝕜` is nontrivially normed.
+-/
+theorem egauge_pi' {I : Set ι} (hI : I.Finite)
+    {U : ∀ i, Set (E i)} (hU : ∀ i ∈ I, Balanced 𝕜 (U i))
+    (x : ∀ i, E i) (hI₀ : I = univ ∨ (∃ i ∈ I, x i ≠ 0) ∨ (𝓝[≠] (0 : 𝕜)).NeBot) :
+    egauge 𝕜 (I.pi U) x = ⨆ i ∈ I, egauge 𝕜 (U i) (x i) := by
+  refine le_antisymm ?_ (iSup₂_le fun i hi ↦ le_egauge_pi hi _ _)
+  refine le_of_forall_lt' fun r hr ↦ ?_
+  have : ∀ i ∈ I, ∃ c : 𝕜, x i ∈ c • U i ∧ ‖c‖ₑ < r := fun i hi ↦
+    egauge_lt_iff.mp <| (le_iSup₂ i hi).trans_lt hr
+  choose! c hc hcr using this
+  obtain ⟨c₀, hc₀, hc₀I, hc₀r⟩ :
+      ∃ c₀ : 𝕜, (c₀ ≠ 0 ∨ I = univ) ∧ (∀ i ∈ I, ‖c i‖ ≤ ‖c₀‖) ∧ ‖c₀‖ₑ < r := by
+    have hr₀ : 0 < r := hr.bot_lt
+    have Hpos [hbot : (𝓝[≠] (0 : 𝕜)).NeBot] : ∃ c₀ ≠ (0 : 𝕜), ‖c₀‖ₑ < r :=
+      frequently_iff_neBot.mpr hbot |>.and_eventually
+        ((continuous_enorm.tendsto' 0 0 (by simp)).eventually_lt_const hr₀) |>.exists
+    rcases I.eq_empty_or_nonempty with rfl | hIne
+    · obtain hι | hbot : IsEmpty ι ∨ (𝓝[≠] (0 : 𝕜)).NeBot := by simpa [@eq_comm _ ∅] using hI₀
+      · use 0
+        simp [@eq_comm _ ∅, hι, hr₀]
+      · rcases Hpos with ⟨c₀, hc₀, hc₀r⟩
+        exact ⟨c₀, .inl hc₀, by simp, hc₀r⟩
+    · obtain ⟨i₀, hi₀I, hc_max⟩ : ∃ i₀ ∈ I, IsMaxOn (‖c ·‖ₑ) I i₀ :=
+        exists_max_image _ (‖c ·‖ₑ) hI hIne
+      by_cases H : c i₀ ≠ 0 ∨ I = univ
+      · exact ⟨c i₀, H, fun i hi ↦ by simpa [enorm] using hc_max hi, hcr _ hi₀I⟩
+      · push_neg at H
+        have hc0 (i : ι) (hi : i ∈ I) : c i = 0 := by simpa [H] using hc_max hi
+        have heg0 (i : ι) (hi : i ∈ I) : x i = 0 :=
+          zero_smul_set_subset (α := 𝕜) (U i) (hc0 i hi ▸ hc i hi)
+        have : (𝓝[≠] (0 : 𝕜)).NeBot := (hI₀.resolve_left H.2).resolve_left (by simpa)
+        rcases Hpos with ⟨c₁, hc₁, hc₁r⟩
+        refine ⟨c₁, .inl hc₁, fun i hi ↦ ?_, hc₁r⟩
+        simp [hc0 i hi]
+  refine egauge_lt_iff.2 ⟨c₀, ?_, hc₀r⟩
+  have smul_pi : c₀ • I.pi U = I.pi (c₀ • U) := by
+    rcases hc₀ with hc₀ | rfl
+    · rw [smul_pi₀ _ _ hc₀]
+    · rw [smul_univ_pi _ _]
```

Comment:
I pushed this small change, which is more lines but fewer lemma invocations. Maybe there's a way to clean this up a little further.

Reply:
(I used the ideas from your old proofs to make #23287)